--- a/src/androidemu/vfs/file_system.py
+++ b/src/androidemu/vfs/file_system.py
@@ -55,8 +55,16 @@ class VirtualFileSystem:
 
         file_path = self._root_path.joinpath(filename).resolve()
 
-        if not file_path.is_relative_to(self._root_path):
-            raise RuntimeError("Emulator tried to read outside vfs ('%s' not in '%s')." % (file_path, self._root_path))
+        # is_relative_to is new in version 3.9
+        if not hasattr(file_path, 'is_relative_to'):
+            try:
+                file_path.relative_to(self._root_path)
+            except:
+                # if file_path is not in the subpath of self._root_path, ValueError is raised
+                raise RuntimeError("Emulator tried to read outside vfs ('%s' not in '%s')." % (file_path, self._root_path))
+        else:
+            if not file_path.is_relative_to(self._root_path):
+                raise RuntimeError("Emulator tried to read outside vfs ('%s' not in '%s')." % (file_path, self._root_path))
 
         return file_path
 
